src/contract/assembled_transaction.ts

```diff
@@ -453,35 +457,82 @@ export class AssembledTransaction<T> {
-          "call `AssembledTransaction.build` first.",
-        );
-      }
+  private static async buildFootprintRestoreTransaction<T>(
+    options: AssembledTransactionOptions<T>,
+    sorobanData: SorobanDataBuilder | xdr.SorobanTransactionData,
+    account: Account,
+    fee: string
+  ): Promise<AssembledTransaction<T>> {
+    const tx = new AssembledTransaction(options);
+    tx.raw = new TransactionBuilder(account, {
+      fee,
+      networkPassphrase: options.networkPassphrase,
+    })
+      .setSorobanData(sorobanData instanceof SorobanDataBuilder ? sorobanData.build() : sorobanData)
+      .addOperation(Operation.restoreFootprint({}))
+      .setTimeout(options.timeoutInSeconds ?? DEFAULT_TIMEOUT);
+    await tx.simulate({ restore: false });
+    return tx;
+  }
 
-      this.built = this.raw.build();
+  simulate = async ({ restore }: {restore?: boolean} = {}): Promise<this> => {
+    if (!this.raw) {
+      throw new Error(
+        "Transaction has not yet been assembled; " +
+        "call `AssembledTransaction.build` first."
+      );
     }
-    this.simulation = await this.server.simulateTransaction(this.built);
+
+    restore = restore ?? this.options.restore;
+    this.built = this.raw.build();
 
     // need to force re-calculation of simulationData for new simulation
     delete this.simulationResult;
     delete this.simulationTransactionData;
+    this.simulation = await this.server.simulateTransaction(this.built);
+
+    if (Api.isSimulationRestore(this.simulation) && restore) {
+      const account = await getAccount(this.options, this.server);
+      const result = await this.restoreFootprint(
+        this.simulation.restorePreamble,
+        account
+      );
+      if (result.status === Api.GetTransactionStatus.SUCCESS) {
+        // need to rebuild the transaction with bumped account sequence number
+        const contract = new Contract(this.options.contractId);
+        this.raw = new TransactionBuilder(account, {
+          fee: this.options.fee ?? BASE_FEE,
+          networkPassphrase: this.options.networkPassphrase,
+        })
+          .addOperation(
+            contract.call(
+              this.options.method,
+              ...(this.options.args ?? [])
+            )
+          )
+          .setTimeout(
+            this.options.timeoutInSeconds ?? DEFAULT_TIMEOUT
+          );
+        await this.simulate();
+        return this;
+      }
+      throw new AssembledTransaction.Errors.RestorationFailure(
+        `Automatic restore failed! You set 'restore: true' but the attempted restore did not work. Result:\n${JSON.stringify(result)}`
+      );
+    }
 
     if (Api.isSimulationSuccess(this.simulation)) {
       this.built = assembleTransaction(
         this.built,
-        this.simulation,
+        this.simulation
       ).build();
     }
 
     return this;
   };
 
   get simulationData(): {
-    result: Api.SimulateHostFunctionResult;
+    result?: Api.SimulateHostFunctionResult;
```

Comment:
Breaking? :sweat:

Reply:
The challenge that @BlaineHeffron was working around here is that, when simulated, the restore transaction ends up having an empty `result` (even though the typing in stellar-sdk indicates this should not be possible!). He originally worked around this a different way, which involved adding a new `updateTimeout` argument to `signAndSend` (see [here](https://github.com/stellar/js-stellar-sdk/pull/974#discussion_r1619287348). You can see our discussion about maybe not cluttering the API and working around it a different way [here](https://github.com/stellar/js-stellar-sdk/pull/974#discussion_r1619287348). That change is what resulted in the `?` being added to `result` here.

It _is_ technically a breaking change, and seems like a necessary one without significant code rework (and adding back that API scope creep that I was hoping to avoid).

Reply:
Another option would be to add a spoof/stub `Api.SimulateHostFunctionResult` in this getter, if the actual simulation didn't return one?

Reply:
ok, that's what I did, which seems like it _might_ introduce a subtle bug, depending on what is expected by `parseResultXdr` 😞